controllers/application_controller.go

```diff
@@ -0,0 +1,349 @@
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Info("finished finalization")
+		return ctrl.Result{}, nil
+	}
+	if argocdApp.GetDeletionTimestamp() != nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	logger.Info("starting finalization")
+	err := r.client.Delete(ctx, argocdApp)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{Requeue: true}, nil
+}
+
+func (r *ApplicationReconciler) reconcileApplication(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	canSync, err := r.canSyncApplicationSpec(ctx, argocdApp, tenantApp)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !canSync {
+		return ctrl.Result{
+			RequeueAfter: 30 * time.Minute,
+		}, nil
+	}
+
+	// Sync application spec from tenant to argocd.
+	err = r.syncApplicationSpec(ctx, argocdApp, tenantApp)
+	if err != nil {
+		logger.Error(err, "failed to sync application spec")
+		return ctrl.Result{}, err
+	}
+	// Sync application status from argocd to tenant.
+	err = r.syncApplicationStatus(ctx, argocdApp, tenantApp)
+	if err != nil {
+		logger.Error(err, "failed to sync application status")
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ApplicationReconciler) canSyncApplicationSpec(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) (bool, error) {
+	logger := log.FromContext(ctx)
+	ns := &corev1.Namespace{}
+	err := r.client.Get(ctx, client.ObjectKey{Name: tenantApp.GetNamespace()}, ns)
+	if err != nil {
+		return false, err
+	}
+	tenantName, ok := ns.Labels[constants.OwnerTenant]
+	if !ok {
+		logger.Info("the namespace does not belong to a tenant", "targetNamespace", ns.Name)
+		r.recorder.Eventf(tenantApp, corev1.EventTypeWarning, "CannotSync", "the namespace '%s' does not belong to a tenant", ns.Name)
+		return false, nil
+	}
+
+	if argocdApp != nil {
+		ownerNs, ok := argocdApp.GetLabels()[constants.OwnerAppNamespace]
+		if !ok {
+			project, found, err := unstructured.NestedString(argocdApp.UnstructuredContent(), "spec", "project")
+			if err != nil {
+				return false, err
+			}
+			if !found {
+				return false, errors.New("spec.project not found in the application: " + argocdApp.GetNamespace() + "/" + argocdApp.GetName())
+			}
+			if project != tenantName {
+				logger.Info("project of the application does not match the tenant name", "project", project, "tenantName", tenantName)
+				r.recorder.Eventf(tenantApp, corev1.EventTypeWarning, "CannotSync", "project '%s' of the application '%s/%s' does not match the tenant name '%s'", project, argocdApp.GetNamespace(), argocdApp.GetName(), tenantName)
+				return false, nil
+			}
+		} else if tenantApp.GetNamespace() != ownerNs {
+			logger.Info("the application is already managed by other namespace", "tenantNamespace", tenantApp.GetNamespace(), "ownerNamespace", ownerNs)
```

Comment:
```suggestion
			logger.Info("the application is already managed by another namespace", "tenantNamespace", tenantApp.GetNamespace(), "ownerNamespace", ownerNs)
```

Please check the following lines for the same issue.

```
controllers/application_controller.go:185
controllers/application_controller.go:186
hooks/application.go:117
hooks/application_test.go:65
hooks/application_test.go:78
```